api/agent/protocol/json.go

```diff
@@ -35,54 +46,92 @@ func writeString(err error, dst io.Writer, str string) error {
 
 // TODO(xxx): headers, query parameters, body - what else should we add to func's payload?
 // TODO(xxx): get rid of request body buffering somehow
-func (h *JSONProtocol) DumpJSON(req *http.Request) error {
+func (h *JSONProtocol) DumpJSON(call *models.Call, req *http.Request) error {
 	stdin := json.NewEncoder(h.in)
 	bb := new(bytes.Buffer)
 	_, err := bb.ReadFrom(req.Body)
+	// todo: better/simpler err handling
 	if err != nil {
 		return err
 	}
+	// open
 	err = writeString(err, h.in, "{")
+	if err != nil {
+		return err
+	}
+
+	// call_id
+	err = writeString(err, h.in, `"call_id":`)
+	if err != nil {
+		return err
+	}
+	err = stdin.Encode(call.ID)
+	if err != nil {
+		return err
+	}
+
+	// body
+	err = writeString(err, h.in, ",")
 	err = writeString(err, h.in, `"body":`)
 	if err != nil {
 		return err
 	}
 	err = stdin.Encode(bb.String())
+	if err != nil {
+		return err
+	}
+
+	// request URL
 	err = writeString(err, h.in, ",")
-	err = writeString(err, h.in, `"headers":`)
+	err = writeString(err, h.in, `"request_url":`)
 	if err != nil {
 		return err
 	}
-	err = stdin.Encode(req.Header)
+	err = stdin.Encode(req.URL.String())
+	if err != nil {
+		return err
+	}
+
+	// headers
 	err = writeString(err, h.in, ",")
-	err = writeString(err, h.in, `"query_parameters":`)
+	err = writeString(err, h.in, `"headers":`)
 	if err != nil {
 		return err
 	}
-	err = stdin.Encode(req.URL.RawQuery)
-	err = writeString(err, h.in, "}")
+	err = stdin.Encode(req.Header)
+
+	// config
+	if call.EnvVars != nil && len(call.EnvVars) > 0 {
+		err = writeString(err, h.in, ",")
+		err = writeString(err, h.in, `"config":`)
+		if err != nil {
+			return err
+		}
+		err = stdin.Encode(call.EnvVars)
+	}
+
+	// close
+	err = writeString(err, h.in, "}\n\n")
 	return err
 }
 
-func (h *JSONProtocol) Dispatch(w io.Writer, req *http.Request) error {
-	err := h.DumpJSON(req)
+func (h *JSONProtocol) Dispatch(call *models.Call, w io.Writer, req *http.Request) error {
+	err := h.DumpJSON(call, req)
 	if err != nil {
 		return err
 	}
-	jout := new(jsonio)
+	jout := new(jsonOut)
 	dec := json.NewDecoder(h.out)
 	if err := dec.Decode(jout); err != nil {
-		return err
+		return fmt.Errorf("error decoding JSON from user function: %v", err)
```

Comment:
this is error handling can be even trickier, what if request body malformed? We might check the body for JSON syntax errors.

Reply:
+1

Reply:
@denismakogon not sure what you mean?  Previously it just showed the parsing error, eg: "Invalid character blah, blah". No context, no nothing to know where this occurred. At least this gives some reference.

Reply:
i believe he meant something along these lines https://github.com/fnproject/fn/pull/424#discussion_r147214118 -- I agree.

Reply:
@rdallman thanks for reffering to that error comment, yes that's what mean.

Reply:
Commented on #424 comment.